Clarify event page/menu registration in MV3 migration guide (#1530)
mozilla/extension-workshop: src/content/documentation/develop/manifest-v3-migration-guide.md

### Event-driven background scripts

Firefox has extended its support for [background scripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts) by enabling non-persistent background pages (aka Event Pages) for Manifest V2 and V3. Using non-persistent background scripts greatly reduces your extension use of browser resources. However, MV3 removes support for persistent background pages.

To migrate your extension to using non-persistent background pages you need to:

- Update your manifest.json `background` key to remove the `"persistent"` property or set it to `false`. This feature is also supported in MV2 since Firefox 106.
- Ensure listeners are at the top-level and use the synchronous pattern.
- Record state changes in local storage.
- Change timers to alarms.
- Switch from using [`extension.getBackgroundPage`](https://developer.mozilla.org/en-US/Mozilla/Add-ons/WebExtensions/API/extension/getBackgroundPage) to call a function from the background page, to [`runtime.getBackgroundPage`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/getBackgroundPage).

> - Place menu creation using [`menus.create`](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/menus/create) or its alias `contextMenus.create` in a `runtime.onInstalled` listener. Also, note that the [`menus.onClicked`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus/onClicked) event or its alias `contextMenus.onClicked` must be used to handle menu entry click events from an event page instead of the `onclick` parameter to the `contextMenus.create` or `contextMenus.update` methods. If the `onclick` property of `menus.create` or its alias `contextMenus.create` are used from a call originating from an event page, they throw synchronously.

Reviewer: ```suggestion
- Place menu creation using [`menus.create`](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/menus/create) or its alias `contextMenus.create` in a `runtime.onInstalled` listener. Also, note that the [`menus.onClicked`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus/onClicked) event or its alias `contextMenus.onClicked` must be used to handle menu entry click events from an event page instead of the `onclick` parameter of the `contextMenus.create` or `contextMenus.update` methods. If the `onclick` property of `menus.create` or its alias `contextMenus.create` are used from a call originating from an event page, they throw synchronously.
```

::: note
Safari also supports event-driven background scripts, however, Chromium has adopted service workers instead.
:::

::: note

> Firefox supports non-persistent background pages since Firefox 106. In Firefox 105 and earlier, background pages were always persistent.

Author: Firefox 105 is in the past, and I wanted to reflect that the state of being unconditionally persistent is past behavior. If "were" is not the right word, would the following alternatives be better?
"were always" -> "had always been" or "used to always be"

> Firefox supports non-persistent background pages since Firefox 106. In Firefox 105 and earlier, background pages were always persistent.

Reviewer: The point is that you are describing current behavior for 105 (and earlier), "were", "had", "used to" create ambiguity suggesting you are describing some earlier behavior in 105 and it could now be different.

> Firefox supports non-persistent background pages since Firefox 106. In Firefox 105 and earlier, background pages were always persistent.

Author: I suppose that the difference is that I'm (erroneously?) using Firefox versions as units of dates, while you're looking at them as snapshots of products (identifiable by a incrementing number).

With the suggested edit, it is not immediately obvious that the old behavior is an artifact of the past, and increasingly more obsolete as time passes (and users update to more recent versions).

Do you have a suggestion to improve that?

> Firefox supports non-persistent background pages since Firefox 106. In Firefox 105 and earlier, background pages were always persistent.

Reviewer: It is because that is what does happen in those versions, not what did happen in those versions (and, therefore, might not happen anymore). How about this
```suggestion
Firefox supports non-persistent background pages from Firefox 106. Firefox supports only persistent background pages in Firefox 105 and earlier.
```

::: note
Safari also supports event-driven background scripts, however, Chromium has adopted service workers instead.
:::

::: note
Firefox supports non-persistent background pages from Firefox 106. In Firefox 105 and earlier, event pages are run as if they are a persistent background page.
:::

More information on the migration process can be found on the [background script](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/Background_scripts) page on MDN.

#### Event pages and backwards-compatibility

::: note
This section is only relevant if your extension supports Firefox 105 and earlier.
:::

> An extension designed as a non-persistent background page generally works even when event pages are not supported (i.e. in Firefox 105 and earlier).
> There is one exception: the registration of context menus. In an event page, context menus persist across restarts, while they do not in persistent background pages.

Reviewer: ```suggestion
An extension designed as a non-persistent background page works even when event pages are not supported (i.e. in Firefox 105 and earlier) 
with one exception: the registration of context menus. In an event page, context menus persist across restarts, while they do not in persistent background pages.
```

If the recommendation to register menus in `runtime.onInstalled` is followed, these menus are removed after a browser restart in Firefox 105 and earlier.
To work around this issue, you could unconditionally call `browser.contextMenus.create`.

> When the menu already exists, the `browser.runtime.lastError` property will be set when the (optional) `create` callback is called.

Reviewer: ```suggestion
When the menu exists, the `browser.runtime.lastError` property is set when the (optional) `create` callback is called.
```
